### ISSUE TITLE
linter(plugins): Improve the error message/config schema for "reserved" plugin names

### ISSUE THREAD

**Issue by connorshea** [oxc-project/oxc#15413]
### What version of Oxlint are you using?

1.26.0

### What command did you run?

`npm run oxlint`

### What does your `.oxlintrc.json` config file look like?

```jsonc
{
  "jsPlugins": ["jsdoc"]
}
```

### What happened?

Error message is returned, `Plugin name 'jsdoc' is reserved, and cannot be used for JS plugins`.

It would be good if we updated the user experience when hitting reserved names:

- Improve this error message (or linked to the docs maybe?), to explain that `jsdoc` rules are already implemented via a native implementation in Rust, and oxlint cannot support it as a jsPlugin because it wouldn't know whether to pull rules you've configured from the native implementation or from JS.
- Add the list of disallowed names to the JSON Schema, so that VS Code/other editors would show a warning in .oxlintrc.json if you put any of the disallowed values into the `jsPlugins` array.

**Comment by overlookmotel** [comment]
@camc314 I've assigned us both because we should discuss.

**Comment by overlookmotel** [comment]
Possible options I can think of:

#### 1. Disallow `jsdoc` being in *both* `jsPlugins` and `plugins`

If it's in `jsPlugins`, all `jsdoc` rules run from the JS plugin only.

Downside: Much slower than in needs to be.

#### 2. Run `jsdoc` rules which are implemented in Oxc as Oxc native rules, and the rest from the JS plugin

Downsides:

* Problematic if Oxc's implementation has a bug - no way for user to switch to the working JS version.
* Problematic if Oxc doesn't implement fixes/suggestions for that rule yet.

#### 3. Completely outlaw `jsdoc` in `jsPlugins` (as now)

User would have to create their own JS plugin `my-jsdoc` which re-exports rules from the original JS `jsdoc` plugin.

---

Note: Any JS plugin can call itself `jsdoc`. There's no guarantee that it's *the* `jsdoc` plugin.

**Comment by overlookmotel** [comment]
Similar to #15258.

**Comment by thernstig** [comment]
A system similar to ESLint flat config would possibly have solved this, as names are not reserved anymore and can be anything. I do hope we get there one day with Oxc as well, but I have not followed the ESLint flat config discussions lately in Oxc. Flat config is discussed here https://github.com/oxc-project/oxc/discussions/4720.

One way to solve this now could have been to allow this, which would be a half-hack of ESLint flat config:

```json
{
  "jsPlugins": [ { "jsdocNative": "jsdoc" }],
  "plugins": [ "jsdoc" ]
  "rules": {
    "jsdocNative/some-rule": "error",
    "jsdoc/some-rule": "off",
  },
}
```

i.e. to support `jsPlugins` to accept an object with a `key: value`, where `key` is a custom name and `value` is the real plugin name.

This way, people can disable rules from `jsdocNative` that are already 100% compliant with the Rust-native `jsdoc` implementation. But [jsdoc moves at a high speed](https://github.com/gajus/eslint-plugin-jsdoc/releases) so even using Rust-native rules could be "dangerous" as @brettz9 does such an incredible job of updating rules and fixing bugs and corner cases.

**Comment by camc314** [comment]
**I asked ChatGPT**

Review https://github.com/oxc-project/oxc/issues/15258 then look at the following two solutions and evaluate them.

Option 1:

```json
{
  "plugins": ["jsdoc"],
  "jsPlugins": ["jsdoc"],
  "rules": {
    "js-plugin/jsdoc/rule-name": "error",
    "jsdoc/rule-name": "error"
  }
}
```

Option 2:

```json
{
  "plugins": [ "jsdoc" ],
  "jsPlugins": [ { "jsdocNative": "jsdoc" } ],
  "rules": {
    "jsdocNative/some-rule": "error",
    "jsdoc/some-rule": "off"
  }
}
```

**it said:**

Short take: pick Option 2 (explicit aliasing). It’s clearer, ESLint-compatible, scales to any collision, and avoids a magic prefix.

Why (grounded in the current situation):
	•	There are real collisions today: e.g. trying to load eslint-plugin-react-hooks or eslint-plugin-jsdoc runs into reserved names used by Oxlint’s native plugins, which is exactly what the issue is about.  ￼
	•	Oxlint’s JS-plugin support aims for ESLint compatibility; forcing users to rewrite every rule id with a special js-plugin/… prefix breaks that expectation and complicates migration. The docs and comms emphasize “ESLint-compatible API.”  ￼

Comparison

Option 1 (special js-plugin/ prefix)
	•	✅ Simple for the engine (no name mapping).
	•	❌ Not ESLint-compatible: users must change all rule ids (jsdoc/* → js-plugin/jsdoc/*).
	•	❌ Leaks an implementation detail (“this rule came from JS”) into config space.
	•	❌ Hurts discoverability and copy-pasting from upstream ESLint configs/blogs.

Option 2 (alias the JS plugin)
	•	✅ ESLint-style: mirrors how flat-config names a plugin namespace distinct from the package name.
	•	✅ Zero surprises: native plugin keeps jsdoc/*; JS plugin can be jsdoc-js/* (or any alias).
	•	✅ Scales to multiple conflicts and coexisting versions (react-hooks-rc, react-hooks-canary, etc.).
	•	✅ Better UX for migration tools (only add an alias when a collision exists).  ￼
	•	⚠️ Slightly richer schema (strings or { alias: package } objects), but worth it.

Recommended shape (tweak of Option 2)

Use a mixed array: strings for simple cases; objects for aliasing conflicts.

```json
{
  "plugins": ["react", "jsdoc"],                         // native Oxlint plugins
  "jsPlugins": [
    "eslint-plugin-import",                              // no conflict, keep shorthand
    { "jsdoc-js": "eslint-plugin-jsdoc" },               // alias to avoid 'jsdoc' collision
    { "react-hooks-rc": "eslint-plugin-react-hooks" }    // alias to avoid 'react-hooks' collision
  ],
  "rules": {
    "jsdoc/some-native-rule": "error",
    "jsdoc-js/some-rule": "error",
    "react/some-native-rule": "warn",
    "react-hooks-rc/unsupported-syntax": "error"
  }
}
```

Nice-to-have engine behavior
	•	Warn (not fail) when a JS plugin tries to use a reserved native name; suggest an alias automatically in the error message. (This is already flagged today; improve the UX with a “try: { ‘name-js’: ‘pkg’ }” hint.)  ￼
	•	If both a native and JS plugin export the same rule id under the same namespace (shouldn’t happen with aliasing), error with a clear conflict message and a fix hint.
	•	Allow a configurable default JS namespace prefix (e.g. js/) but don’t require it; aliasing remains the primary path.

Bottom line
	•	Option 2 preserves ESLint mental models, keeps native rule ids stable, and offers precise, explicit control when you need both native and JS variants side-by-side—exactly what your react-hooks/jsdoc scenarios need.  ￼

**Comment by zoontek** [comment]
@camc314 ChatGPT is hallucinating, there's no way to create plugin aliases like this.

**Comment by camc314** [comment]
sorry let me clarify. that isn't a fix to the issue, but rather an approach to how we're going to fix the issue. me and @overlookmotel were pairing on a call this morning

**Comment by thernstig** [comment]
Glad ChatGPT agrees with my suggestion then 😆

**Comment by overlookmotel** [comment]
> Glad ChatGPT agrees with my suggestion then 😆

If it's good enough for GPT, it's good enough for me! That's how we make decisions around here. 🤡

No, Cam and I both like your idea @thernstig. It's an elegant solution which (I think) should cover all edge cases.

## Possible variants

I also wonder about a few potential variants:

#### Option 3:

Same as @thernstig's suggestion, but a little more flexible. Allow (but don't require) multiple "aliased" plugins to be specified in a single object.

```ts
type JsPlugins = (string | Record<string, string>)[];
```

```json
{
  "jsPlugins": [
    "eslint-plugin-import",
    {
      "jsdoc-js": "eslint-plugin-jsdoc",
      "react-hooks-rc": "eslint-plugin-react-hooks"
    },
  ]
}
```

#### Option 4:

More verbose, but more explicit.

```ts
type JsPlugins = (string | { name: string, path: string })[];
```

```json
{
  "jsPlugins": [
    "eslint-plugin-import",
    { "name": "jsdoc-js", "path": "eslint-plugin-jsdoc" },
    { "name": "react-hooks-rc", "path": "eslint-plugin-react-hooks" }
  ]
}
```

#### Option 5:

`jsPlugins` is an object. More flat config-like. But the `: true` variant is ugly.

```ts
type JsPlugins = { [name: string]: string | true };
```

```json
{
  "jsPlugins": {
    "eslint-plugin-import": true, // Shorthand for default name
    "jsdoc-js": "eslint-plugin-jsdoc",
    "react-hooks-rc": "eslint-plugin-react-hooks"
  }
}
```

#### Option 6:

`jsPlugins` is an object. All plugins must be explicitly named. Most flat config-like.

```ts
type JsPlugins = { [name: string]: string };
```

```json
{
  "jsPlugins": {
    "import": "eslint-plugin-import",
    "jsdoc-js": "eslint-plugin-jsdoc",
    "react-hooks-rc": "eslint-plugin-react-hooks"
  }
}
```

### My thoughts

I'm not pushing any of these alternatives, just exploring possibilities.

Personally I don't like the last 2 (the flat config-like ones) because they're optimized for the uncommon case that you want to alias the plugin name.

Any thoughts?

## Default name

This also raises the question of what default name for plugins should be when no alias is provided.

Since we're now making the config (at least partially) responsible for naming, where should we get the default name from?

1. Plugin path as specified in config.
2. From the plugin's `meta.name` property.
3. From the plugin package's name (i.e. `name` property in `package.json`).

The first option is less ideal for custom plugins which aren't an NPM package. e.g.:

```json
{
  "jsPlugins": [
    "../../plugins/my-plugin/index.js",
    "../../plugins/my-other-plugin/index.js"
  ]
}
```

We could sidestep that issue by requiring that plugins specified by path must have an alias provided.

What does ESLint do? 🤔

### When Oxlint adds a rule natively

When Oxlint implements a rule natively which user is currently using from a JS plugin, how do we help/prompt the user to switch to the native version, to get the benefit of the faster implementation?

The one advantage (and peril!) of [my previous suggestion](https://github.com/oxc-project/oxc/issues/15413#issuecomment-3508978193) is that this would happen automatically:

> 2. Run jsdoc rules which are implemented in Oxc as Oxc native rules, and the rest from the JS plugin

That solution has other problems, but we should try to find a way to encourage users to switch to native versions of rules when they're available.

**Comment by thernstig** [comment]
Immediate reaction: your **option 3** looks best.

For where to get the name I _believe_ `meta.name` is the most proper, but I wager this will be discovered and understood more when trying to implement a prototype, so it could change (naturally).

> When Oxlint implements a rule natively which user is currently using from a JS plugin, how do we help/prompt the user to switch to the native version, to get the benefit of the faster implementation?

This one is hard.

If we look at my favorite https://github.com/gajus/eslint-plugin-jsdoc, new options are added to existing rules very often. There is thus a danger to rely on just checking rule names and say "please switch to Oxc-native, it is faster". I am unsure what the solution is here. I suspect you cannot in any sensible way check if your Oxc implementations behave exactly the same as the equivalent ESLint implementation for all corner cases like rule semantic changes, new options etc? Seems "hard".

To be honest, if this is impossible, and the stance of Oxc is to at least support the most common plugins, I think Oxc should deprecate fast-moving plugins (such as JSDoc) and rely on ESLint native ones. And then make absolutely sure that the Oxc-implemented ones has a cadence to track their upstream counterparts to always implement them one-to-one. Would be a full-time job of a person most likely though. But that is the only sensible long-term solution I can think of.

**Comment by connorshea** [comment]
Need to also consider the case where there are rules which didn't get ported because they have type-aware equivalents, and whether we risk leading users into confusing behaviors/redundant rules fighting each other if this isn't handled elegantly.

**Comment by connorshea** [comment]
Rather than allowing an arbritrary alias, what if we force an alias? Like `js@identifier` so jsdoc becomes `js@jsdoc`, and so on, based on whatever the ESLint plugin calls itself. Otherwise I imagine it could become difficult to figure out where any given rule is coming from, and just generally may make passing rules around more difficult.

I think arbitrary aliases can also cause problems in the future if oxlint ever gains things like community-maintained shareable configs.

Only problem is that @ is already usable in plugin names :s 

Also, if we don't do the above, we should probably also disallow aliasing the same plugin multiple times with different names via overrides.

Would it be crazy to _only_ allow aliases for plugins with the same name as those that are native to oxlint?

**Comment by thernstig** [comment]
@connorshea that might work, but there are real reasons why ESLint flat config uses aliases. Unsure if they apply here though, but it is a good way to massage people into learning them if Oxc ever goes that route of using a flat config (I hope it does).

https://eslint.org/blog/2022/08/new-config-system-part-1/
